call.go

```diff
@@ -29,14 +29,27 @@ import (
 func (cc *ClientConn) Invoke(ctx context.Context, method string, args, reply interface{}, opts ...CallOption) error {
 	// allow interceptor to see all applicable call options, which means those
 	// configured as defaults from dial option as well as per-call options
-	opts = append(cc.dopts.callOptions, opts...)
+	opts = combine(cc.dopts.callOptions, opts)
 
 	if cc.dopts.unaryInt != nil {
 		return cc.dopts.unaryInt(ctx, method, args, reply, cc, invoke, opts...)
 	}
 	return invoke(ctx, method, args, reply, cc, opts...)
 }
 
+func combine(o1 []CallOption, o2 []CallOption) []CallOption {
+	// we don't use append because o1 could have extra capacity whose
+	// elements would be overwritten, which could cause inadvertent
+	// sharing (and race connditions) between concurrent calls
+	if len(o1) == 0 && len(o2) == 0 {
+		return nil
+	}
```

Comment:
To optimize further, how about this instead:

```go
if len(o1) == 0 {
  return o2
} else if len(o2) == 0 {
  return o1
}
```

Reply:
Good idea. Done.